[16.0][IMP] stock_barcodes

edescalona opened the pull request:
@BinhexTeam The different views for the barcode have been modified.

rousseldenis commented:
@edescalona As the changes in this are quite different than the PR description, could you enhance it to enumerate them?

And it is better to add the module name in the commit messages like `[IMP] stock_barcodes: (...)`.

Thanks!

edescalona commented:
> @edescalona As the changes in this are quite different than the PR description, could you enhance it to enumerate them?
> 
> And it is better to add the module name in the commit messages like `[IMP] stock_barcodes: (...)`.
> 
> Thanks!

Hello bruo, ok, I'll upload other improvements in a while and with their description, greetings

edescalona commented:
> @edescalona As the changes in this are quite different than the PR description, could you enhance it to enumerate them?
> 
> And it is better to add the module name in the commit messages like `[IMP] stock_barcodes: (...)`.
> 
> Thanks!

Hello @rousseldenis , ok, I'll upload other improvements in a while and with their description, greetings

mrsavage commented:
I really would like to have this :) Thanks for the work

edescalona commented:
Hi @mrsavage  @rousseldenis , the PR is ready to be reviewed again, any opinion is welcome, thanks.

christian-ramos-tecnativa commented:
@edescalona Can you check the files changed? There are modules that shouldn't be envolved here

edescalona committed:
[IMP] stock_barcodes: Visual and functional improvements to the Inventory and Operations scan views

[IMP] stock_barcodes: Updating menu with the new action.

[IMP] stock_barcodes: If the required quantity of the product is exceeded, allow the user to decide on the excess quantity.

[IMP] stock_barcodes: Change format string

[IMP] stock_barcodes: Readme

[IMP] stock_barcodes: Readme

[FIX] stock_barcodes: Validating existing method

[IMP] stock_barcodes: Changing the format of f-strings (f) to {}.format()

[IMP] stock_barcodes: Add bucle for in constrains. Add return domain

[IMP] stock_barcodes: Changing ID of barcode main action

[IMP] stock_barcodes: Add tests.

[IMP] stock_barcodes: Showing the pickings of the same type as the one previously validated.

[IMP] stock_barcodes: Updating menu with the new action.

[IMP] stock_barcodes: Add styles button confirm force done.

edescalona commented:
> @edescalona Can you check the files changed? There are modules that shouldn't be envolved here

Hi @Christian-RB  I have already checked and corrected the changes in other modules that should not be involved. Ready to be reviewed, thanks for your contribution.

hitrosol commented:
Thanks for this PR.
I try this on runboat. 
1. Set the Barcode Option Group  of Picking IN options in the Receipts
2. Click the Picking IN, in the Barcode menu.
3. Click the x to process button. 
4. It shows the Picking In orders. 
5. Validate one of them.
6. The screen now show all picking order. IMHO it should only shows the same picking type order.

edescalona commented:
> Thanks for this PR. I try this on runboat.
> 
> 1. Set the Barcode Option Group  of Picking IN options in the Receipts
> 2. Click the Picking IN, in the Barcode menu.
> 3. Click the x to process button.
> 4. It shows the Picking In orders.
> 5. Validate one of them.
> 6. The screen now show all picking order. IMHO it should only shows the same picking type order.

Hello @hitrosol , thank you for your contribution. I will review the details you mentioned. Thank you very much.

edescalona commented:
Hello @hitrosol  I have already made the change you suggested, thanks for your contribution, it can be reviewed again, greetings

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,117 @@ class StockBarcodesAction(models.Model):
+    count_elements = fields.Integer(default=0, compute="_compute_count_elements")
+
+    @api.constrains("barcode")
+    def _constrains_barcode(self):
```
When using "constrains" we usually should use a for loop, since self can have multiple objects

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,117 @@ class StockBarcodesAction(models.Model):
+        if not re.match(REGEX.get("barcode", False), self.barcode):
+            raise ValidationError(
+                _(
+                    f"The barcode {self.barcode} is not correct. "
```
Odoos translation engine (babel) doesn't support F-strings in the used version. 
Please use format strings or something similar.

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,117 @@ class StockBarcodesAction(models.Model):
+        if len(matched_actions) > len(all_barcode):
+            raise ValidationError(
+                _(
+                    f"Barcode has already been assigned to "
```
Odoos translation engine (babel) doesn't support F-strings in the used version.
Please use format strings or something similar.

edescalona replied:
Your comment has been resolved, thank you for your contribution.

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,117 @@ class StockBarcodesAction(models.Model):
+
+    @api.constrains("context")
+    def _constrains_context(self):
+        if self.context and not bool(
```
Odoos translation engine (babel) doesn't support F-strings in the used version.
Please use format strings or something similar.

edescalona replied:
Your comment has been resolved, thank you for your contribution.

baguenth commented on stock_barcodes/static/description/index.html:
I guess this file should not be modified in the PR. From my knowledge this will be generated automatically.

baguenth commented on stock_barcodes/README.rst:
I guess this file should not be modified in the PR. From my knowledge this will be generated automatically.

baguenth commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -396,7 +408,7 @@ def _prepare_stock_moves_domain(self):
         if self.picking_id:
             domain.append(("picking_id", "=", self.picking_id.id))
-        return domain
+        return []
```
This function statically returns [] now. I guess this is not correct.

edescalona replied:
Your comment has been resolved, thank you for your contribution.

baguenth reviewed (changes requested):
Your pull requests could have more code documentation from my point of view. 
Currently there are only a few doc strings and code comments. 

Especially when changing a lot of code (what you did) I highly recommend to document the code well. This helps reviewers to understand the changes.

edescalona commented:
Hi @baguenth  I've already reviewed your comments, thanks for your contribution

edescalona commented:
> Your pull requests could have more code documentation from my point of view. Currently there are only a few doc strings and code comments.
> 
> Especially when changing a lot of code (what you did) I highly recommend to document the code well. This helps reviewers to understand the changes.

Ok, I will add some other indications for better understanding, thanks

edescalona commented:
Hi @baguenth , I have already made improvements to the documentation. Please let me know when you can. Thanks for your contribution.

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,123 @@ class StockBarcodesAction(models.Model):
+        for action in self:
+            if not re.match(REGEX.get("barcode", False), action.barcode):
+                raise ValidationError(
+                    _(
```
Tbh I am not sure if we can and should use the format part within the translation bracket. Better use:

_("...").format(...)

baguenth commented on stock_barcodes/models/stock_barcodes_action.py:
```diff
@@ -19,6 +33,123 @@ class StockBarcodesAction(models.Model):
+            matched_actions = self.sudo().search(domain, order="id")
+            if len(matched_actions) > len(all_barcode):
+                raise ValidationError(
+                    _(
```
Tbh I am not sure if we can and should use the format part within the translation bracket. Better use:

_("...").format(...)

edescalona commented:
Hi @baguenth , you can now review the changes, thanks for your contribution.

pedrobaeza reviewed (dismissed):
I put a preventive requested changes until @sergio-teruel checks this.

andreampiovesana commented:
@edescalona can you fix to merge?

edescalona commented:
Hello @andreampiovesana , the change requests have been resolved. We are awaiting review from the other person referenced in the previous message.

edescalona commented:
Hi @pedrobaeza @sergio-teruel  @ws were you able to review this PR, thanks

andreampiovesana reviewed (approved):
LGTM

christian-ramos-tecnativa commented on stock_barcodes/views/stock_barcodes_action_view.xml:
```diff
@@ -1,36 +1,100 @@
+    </record>
+
+
+    <record id="action_stock_barcodes_action" model="ir.actions.client">
+        <field name="name">Barcodes</field>
+        <field name="tag">stock_barcodes_main_menu</field>
+        <field name="target">fullscreen</field>
+    </record>
```
@edescalona Reusing this id will require a migration script on this version, unless you use a different id, not sure how hard would that be
`For external id stock_barcodes.action_stock_barcodes_action when trying to create/update a record of model ir.actions.client found record of different model ir.actions.act_window`

edescalona replied:
Ok @Christian-RB , I'll check it and tell you, thanks.

rousseldenis commented:
@sergio-teruel

sergio-teruel reviewed (changes requested):
Sorry @edescalona I try to install on a customer in v16 and I think that the original functionality is broken... 
So I can not accept this PR...

Could we have a meet and comment abou it?

Maybe tomorrow??

sergio-teruel commented:
hi everyone, @edescalona @Christian-RB and me met and review the new styles and workflows and seems it's Ok... Althoug we will have to do some changes in other modules...

It's a great contribution!!

christian-ramos-tecnativa reviewed (approved):
LGTM, let me know @sergio-teruel if you have some time to check it. Thanks in advance

cvinh commented:
cc @vehi-invitu can you do a functional review with this PR please ?

mktsrl commented:
Hi @edescalona and everyone involved in this PR,
 
First of all, thank you for this really interesting development!
 
Following the functional review request, I’ve been running some preliminary tests on PR #668 (Runbot and our local Odoo 16 CE instance). It's clear this is a "work in progress," with frequent updates. In fact, some issues I had encountered last week have disappeared, or maybe new ones are emerging. Currently the main issues I am seeing are related to:
- JavaScript (which for us fails in many situations)
- some filter logic
- barcode encoding of movements

I have collected some results that I could document  (e.g., steps to reproduce potential errors) if it would be useful at this stage. I am aware that the PR is a draft and under active development.
 
Let me know if providing this feedback now would be helpful to you and if I can do it via PR comment.
 
Thanks!

pedrobaeza commented:
Yes, I think they are useful in this context before merging this.

edescalona commented:
hi @mktsrl , yes of course all comments are welcome, thanks

mktsrl commented:
Hi @escalona and everyone involved,
as previously mentioned, I report the results of my tests below.
Testing was done on Runbot (commit e374b24) on 06/11/2025, focusing on integrating barcode scanning into the warehouse shipping flow originating from a sales order.

**NOTE**: It is possible that some of the behaviors encountered are due to incorrect usage on my part, especially considering the state of the PR in progress

**Test Scenario:**

1.  Created a sales order for two packages, each containing 10 units of a product (total 20 units, available in stock).
2.  Confirmed the sales order, which generated a delivery `stock.picking`. This picking included a single operation line for both packages.
3.  Printed the delivery slip/picking document, which displays the picking's barcode: 
![1](https://github.com/user-attachments/assets/4046c033-3195-43ae-93b1-ac1436149047)

When accessing the barcode interface through different methods, I encountered the following:

**Observations and Potential Issues:**

*   **[Issue 1: Picking Barcode Encoding/Format in Search]**
    When navigating via *Picking OUT > Delivery Orders* (or similar paths to select a picking), if I try to find the specific picking (in my case, `WH/OUT/00028`) by scanning the barcode from the printed slip (see `[IMAGE 1.png]`), the search fails. It appears the system expects/encodes the picking barcode with hyphens (`-`) instead of slashes (`/`) (e.g., the interface shows `WH-OUT-00028` when the scanned barcode and picking reference is `WH/OUT/00028`).
![1-bis](https://github.com/user-attachments/assets/939313da-6a62-44dd-b1c6-54e000976e37) 
![2](https://github.com/user-attachments/assets/790eadde-49b7-4558-97af-e26eb41e66d3)

*   **[Issue 2: "Operations" Search via Product/Package Barcode]**
    From the main barcode interface, if I click the "Operations" button and then the barcode scan icon, scanning either a package barcode or a product barcode yields no results (the interface remains empty).
![3](https://github.com/user-attachments/assets/d0162331-957a-4cfc-b431-5b147bc82c7c)
![4](https://github.com/user-attachments/assets/44b08257-6c09-4788-ac56-1db8b7c8f188)

*   **[Issue 3: Picking Search via Product/Package Barcode]**
    From the main barcode interface, if I click one of the flow buttons (e.g., "Picking OUT"), select the operation type, the list of ready pickings is displayed. If I then try to scan a product or package barcode as suggested by the dark bar ("Scan a transfer or a product to filter your records"), the list is not filtered, and the interface remains unchanged or empty.
![5](https://github.com/user-attachments/assets/4f8d12f2-d8de-4b34-a67f-f12093a6b382)
![5-bis](https://github.com/user-attachments/assets/87c9345a-cf82-45e1-ac01-166d610be199)

**Workaround and Functional Flow:**

To proceed with testing, I bypassed the search issues above by accessing the picking `WH/OUT/00028` either by:
    a. Using the standard back-office interface (manually filtering the picking list).
    b. Navigating from the sales order to its related Delivery.
Using the "Scan Barcode" Smart Button directly from the picking record successfully opens the barcode interface for the desired picking.

On this interface (opened via the Smart Button), I noted the following:

*   **[Issue 4a: Missing Quantity Indicator in "Operations" Detail]**
    In the screen listing the picking's "detailed operations," after scanning a package barcode, there's no clear counter displayed for processed units versus total units (e.g., "done quantity / reserved quantity") for the just-scanned package line.
![6](https://github.com/user-attachments/assets/b94f69cd-888c-4d9d-8d45-72960d1f86ec)

*   **[Issue 4b: Quantity Indicator Not Updating/Error in Operation Edit Mode]**
    When accessing the detail of an operation line (by clicking to edit it), if a package or product barcode is scanned:
    1.  The quantity text indicator (e.g., "0 / 20 Units") does not increment.
    2.  The system displays a "Barcode not found with this screen values" message.
    However, if I then click the button to add the maximum quantity (e.g., the "+20" button), the "Done" quantity updates correctly, and the system *then* appears to correctly recognize the product barcode it previously reported as not found.
![7](https://github.com/user-attachments/assets/e65b3fae-8d60-41aa-a9d4-5c895c935cfa)
![8](https://github.com/user-attachments/assets/72b6ecbf-cfc1-4649-8377-38f9b6fcd20c)
![9](https://github.com/user-attachments/assets/ef1aeb31-85ae-4e8f-abbc-6d61c3ad96cb)
![10](https://github.com/user-attachments/assets/e26b8b91-948a-4212-8648-1115f99262b5)
![11](https://github.com/user-attachments/assets/d105b3a8-7a01-41b0-bf65-76a77952ed85)

**Notes on Other Errors (JavaScript):**

Regarding previous comments where I mentioned JavaScript errors:

*   **On Runbot:** The `UncaughtClientError > TypeError: args.emitter.rootRef.el is null` error, which previously occurred (e.g., when clicking main menu buttons in the barcode interface as shown in ![12](https://github.com/user-attachments/assets/fb67256d-26c4-4862-b437-3b4cd9a86464) **appears to be RESOLVED**. I no longer encounter it in these latest tests on Runbot.

*   **On Local Installation:** However, this specific JavaScript error (`TypeError: args.emitter.rootRef.el is null` with the following stack trace) **PERSISTS** when I try to install and test this PR on my local Odoo 16.0 CE installation (with OCB updated to today's date).
    ```
    UncaughtClientError > TypeError
    Errore Javascript non rilevato > args.emitter.rootRef.el is null
    TypeError: args.emitter.rootRef.el is null
    onDropdownStateChanged@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:577:4
    setup/<@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:565:1061
    trigger@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:1276:63
    changeStateAndNotify@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:571:118
    close@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:572:21
    setup/<@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:14505:59
    wrapCallback/<@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:2163:93
    trigger@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:1276:63
    onMounted@http://127.0.0.1:8069/web/assets/240-2cf2ef7/web.assets_backend.min.js:5943:19
    wrapError/<@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:1614:506
    complete@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:1513:165
    processFiber@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:2101:43
    processTasks@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:2095:62
    flush/this.frame<@http://127.0.0.1:8069/web/assets/226-cf2d506/web.assets_common.min.js:2092:67 
    ```

I hope these observations are helpful.
As mentioned, it's quite possible that some of these "quirks" are simply down to my own interpretation of the intended workflow for this new interface not being quite perfect yet – if so, my apologies in advance for any "noise"!

Thanks!

edescalona commented:
Hi @mktsrl , thanks for your comments, I'll check them out and let you know.

edescalona commented:
Hi @mktsrl , I've tried the cases you mentioned and they're working for me. However, I'm going to try the same barcodes you show as is. I'm not sure why you're getting the JavaScript error. What you're showing is a core error when activating or deactivating a component. Should I try clearing the browser cache? Thanks again for your feedback. We'll stay in touch.

edescalona commented:
Hi @mktsrl , are you setting the Barcode options according to what you want to do?

mktsrl commented:
I think and hope I've done everything I needed, even though I didn't understand very well how to configure the options from the documentation (the readme also makes an error with the link to the images). However, on runbot I left the configurations that were already present. Instead, I'll send you the screen shot of the configuration I have for "picking out" on my local development machine.
![image](https://github.com/user-attachments/assets/74292934-e6f0-4dac-8f85-e36777b9e506)

TheerayutEncoder commented:
Dear @edescalona 

First of all, thank for your work, this is great work after I have tried this module on Odoo 15
on version 16 there are better UX/UI

I just faced some issue, it look like I could scan barcode actions
This is my step
1. Print out some barcode action.
<img width="1122" height="370" alt="Selection_597" src="https://github.com/user-attachments/assets/f6e2f412-2de0-4794-bcdf-7470fe94a8a8" />

2. Go to *barcode menu* and then try to scan each action but it nothing happen
<img width="1591" height="835" alt="Selection_596" src="https://github.com/user-attachments/assets/efa30daf-1141-42ad-b5c6-43852f2917a1" />

Bilbonet commented on stock_barcodes/__manifest__.py:
Could you include _python-barcode_ dependency?

pedrobaeza replied:
Originally the module didn't need that dependency AFAIK.

edescalona replied:
> Could you include _python-barcode_ dependency?

Hi @Bilbonet , exactly as @pedrobaeza  says, the original module does not have that dependency, thanks for your comments.

TumbaoJu commented:
@pedrobaeza : I am preparing a talk for the OCA Days about these improvements.  Do you know if this PR will be merged by then or should I point to this PR in my talk?

pedrobaeza commented:
AFAIK, the minimum previous features were not met, but @sergio-teruel should confirm.

christian-ramos-tecnativa commented:
@pedrobaeza The last comment from @sergio-teruel https://github.com/OCA/stock-logistics-barcode/pull/668#issuecomment-2916701906 was positive about this change. We could meet up again if needed.

pedrobaeza commented:
I'm asking for confirmation. He's on vacation though. Please give me a day.

pedrobaeza commented:
Thanks all for the work.

/ocabot merge major

houzefa-abba commented:
Hi @edescalona thank you for improving this excellent module! :heart_eyes_cat: 
I took a quick glance via runboat, the new interface looks more mobile/scanner-oriented, very nice.

I noted minor issues https://github.com/OCA/stock-logistics-barcode/issues/703 & https://github.com/OCA/stock-logistics-barcode/issues/704

Before I update this module in existing projects, I would like to know if we have some changelog and especially what breaking changes to look out for.
[The commit message](https://github.com/OCA/stock-logistics-barcode/commit/0114cb26e902ef472d2b721d2e029cb938f84648) has a global list which is helpful to start with; I wonder if we have a more detailed set of changes.
Though maybe I can wait for the talk @TumbaoJu is working on?

Thanks

TumbaoJu commented:
> Hi @edescalona thank you for improving this excellent module! 😻 I took a quick glance via runboat, the new interface looks more mobile/scanner-oriented, very nice.
> 
> I noted minor issues #703 & #704
> 
> Before I update this module in existing projects, I would like to know if we have some changelog and especially what breaking changes to look out for. [The commit message](https://github.com/OCA/stock-logistics-barcode/commit/0114cb26e902ef472d2b721d2e029cb938f84648) has a global list which is helpful to start with; I wonder if we have a more detailed set of changes. Though maybe I can wait for the talk @TumbaoJu is working on?
> 
> Thanks

Hello @edescalona , my talk will be more functional than technical but I will have someone technical with me who maybe would be able to answer your questions at the end of the talk.

etobella commented:
@edescalona You added a dependancy on the module :

python-barcode

You should have included it on the external_dependancies

pedrobaeza commented:
And why that dependency? I already warned you about it on the comments. I think it should be removed unless strong reasons are told.

etobella commented:
@edescalona told us diferently....

https://github.com/OCA/stock-logistics-barcode/pull/668#discussion_r2279068026

edescalona commented:
Hi @etobella , I'll check out that dependency and comment. I think something extra was added. I'll confirm and create PR.

etobella commented:
You don't need to check, it was added a dependency on the module:

`python-barcode`

Someone raised the comment https://github.com/OCA/stock-logistics-barcode/pull/668#discussion_r2278654529

Your answer was "the original module does not have that dependency, thanks for your comments." That  is true, but you added a dependency. It should be avoided in stable modules. Why? Easy, it made it break several CI of other people as it required a library that was not declared and then, impossible to detect by PyPi, and also, a problem if you pin libraries.

The CI of the repo was working because another module had the dependency and then, the library was installed, but that is obviously luck.

edescalona commented:
@etobella @pedrobaeza @Christian-RB Reviewing in detail whether that barcode library is used to generate the barcode image that is added to the barcode actions, and thus be able to download said image.

<img width="1914" height="292" alt="image" src="https://github.com/user-attachments/assets/5599e8d0-0491-49de-83fa-c49966c40076" />

This makes it possible to scan a barcode and be directed directly to the desired action.

edescalona commented:
Ok @etobella , sorry for the inconvenience, should I make a PR to add that dependency to the manifest?

etobella commented:
And that can make it fail other instances. Can you find a solution without this library? Thanks.

Odoo is generating barcodes with other libraries, so probably this is not needed.

edescalona commented:
Ok @etobella , I'll look for another option, I'll create a PR to remove that dependency.

etobella commented:
@edescalona This is the fix: https://github.com/OCA/stock-logistics-barcode/pull/713

edescalona commented:
Exactly @etobella, I'm adjusting the report. It's a simple change. I'll ping you when I'm done. Also, to show you the barcode with `humanreadable` in the method I use. ok?

etobella commented:
We already did the fix

edescalona commented:
ok @etobella thanks
